### PR TITLE
[FVM] update fvm state to support child/merge

### PR DESCRIPTION
**Pull request by ramtinms**
This PR
- updates the state with proper nested child/merge functionality 
- adds a stateManager that controls which state is active.
- merges the uuid functionality into a single state service 
- updates all of the state services (accounts, uuid gen, address gen) to not hold states and accepts a state manager. (with the future plan to make them singular)

### CODE REVIEW

File: fvm/state/state.go
Comment (m4ksio) [inline fvm/state/state.go:17]: I'm not sure if we need a hierarchy of states. State works as a wrapper for view with extra operation which only apply on one level (like fees)
We should probably just keep `delta.View` hierarchy here

**Review by m4ksio — APPROVED** [review]

File: fvm/state/state_manager.go
Comment (janezpodhostnik) [inline fvm/state/state_manager.go:47]: If I understand correctly `RollUp(false, true)` would be used, because the verification nodes still need to confirm this transaction failed.

however `RollUp(true, false)` is not does not seem like it would be ever valid

File: fvm/state/state_manager.go
Comment (janezpodhostnik) [inline fvm/state/state_manager.go:58]: could be merged into `else if`

File: fvm/state/state.go
Comment (janezpodhostnik) [inline fvm/state/state.go:135]: Could (should?) be private

File: fvm/state/state.go
Comment (janezpodhostnik) [inline fvm/state/state.go:274]: maybe we could check that s.parent is nil here. This should only be done on the root right?

File: fvm/state/state.go
Comment (janezpodhostnik) [inline fvm/state/state.go:190]: I expected the write counter to be incremented here... But now I'm not sure how the write counter works.

As a separate question: should the write counter be incremented twice if we are writhing twice to the same key.

File: fvm/state/state.go
Comment (janezpodhostnik) [inline fvm/state/state.go:253]: Ok... I think I understand how the write counter works now... and is this line necessary?

**Review by janezpodhostnik — COMMENTED** [review]
This is huge change! nice! Left a few comments.

File: fvm/state/state_manager.go
Comment (ramtinms) [reply fvm/state/state_manager.go:47]: correct, so might actually  break this method into RollUpWithMerge and RollUpOnlyTouch

File: fvm/state/state.go
Comment (ramtinms) [reply fvm/state/state.go:253]: this one is necessary but toBeWritten won't be updated since the delta are not merged

File: fvm/state/state_manager.go
Comment (ramtinms) [reply fvm/state/state_manager.go]: updated at d63488c

Comment (m4ksio) [reply fvm/state/state_manager.go]: This description of the method looks extremely complex. Are we 100% sure and certain such a complex logic is the only way to do, what in fact, is quite easy (conceptually) changeset management.
I'm trying to raise it now, because its easier to step back and maybe defer to using `View` structures now instead of maybe refactoring it later

**Comment by ramtinms** [comment]
I update this PR to use an injected view for FVM to be able to reuse delta.View functionality inside the state

File: fvm/state/state.go
Comment (ramtinms) [reply fvm/state/state.go:182]: I removed the point of confusion, WriteCounter will be what you need.

File: fvm/state/state.go
Comment (ramtinms) [reply fvm/state/state.go]: I removed this logic completely so we don't need this function

**Review by turbolent — COMMENTED** [review]
Amazing work @ramtinms!

This is quite a large PR, is there any way to split it into smaller ones to make it easier to review it? Or maybe we can have a session to go over it together?

File: fvm/state/state.go
Comment (ramtinms) [reply fvm/state/state.go]: fixed.

File: fvm/state/accounts.go
Comment (ramtinms) [inline fvm/state/accounts.go:515]: added these to removes dependencies, going to move them to the utils

**Review by m4ksio — APPROVED** [review]

File: fvm/utils/test.go
Comment (yhassanzadeh13) [inline fvm/utils/test.go:43]: Just as a heads up this implementation seems **non-atomic**, i.e., if an error occurs it leaves the merge incomplete on the _original_ ledger. Maybe as a test helper, it is not a critical constraint to be atomic.

File: fvm/state/view.go
Comment (janezpodhostnik) [inline fvm/state/view.go:10]: I like this naming better!

File: fvm/state/state_holder.go
Comment (janezpodhostnik) [inline fvm/state/state_holder.go:11]: This isn't used anywhere, probably could be removed.

File: fvm/transactionStorageLimiter.go
Comment (janezpodhostnik) [inline fvm/transactionStorageLimiter.go:26]: how come the storage limiter (and transaction fee deductor) doesn't have a block like this:
```go
	parentState := sth.State()
	childState := sth.NewChild()
	defer func() {
		if mergeError := parentState.MergeState(childState); mergeError != nil {
			panic(mergeError)
		}
		sth.SetActiveState(parentState)
	}()
```

File: fvm/utils/test.go
Comment (yhassanzadeh13) [inline fvm/utils/test.go:98]: shouldn't this method be pointer receiver?

File: fvm/utils/test.go
Comment (yhassanzadeh13) [inline fvm/utils/test.go:123]: Also this method on being pointer receiver.

File: fvm/transactionStorageLimiter.go
Comment (ramtinms) [reply fvm/transactionStorageLimiter.go:26]: I think the storage limiter requires a more complex logic, if storage fails we need to even drop the delta of the previous step. so I didn't add it for now.

Comment (janezpodhostnik) [reply fvm/transactionStorageLimiter.go:26]: Good point. I actually have plans for that! I will show you on Monday.

**Review by yhassanzadeh13 — COMMENTED** [review]
I am still reviewing this PR, amazing work! but takes a while. In the mean time I post my comments.

File: fvm/utils/test.go
Comment (ramtinms) [reply fvm/utils/test.go:59]: actually thats part of the interface design, we still need to update metrics and interactions like read counter so we need it to be pointer.

File: cmd/util/ledger/migrations/accounts.go
Comment (yhassanzadeh13) [inline cmd/util/ledger/migrations/accounts.go:137]: minor comment but these member variables better being unexportable.

File: cmd/util/ledger/migrations/accounts.go
Comment (yhassanzadeh13) [inline cmd/util/ledger/migrations/accounts.go:140]: maybe considering passing this by reference.

Comment (yhassanzadeh13) [inline cmd/util/ledger/migrations/accounts.go:140]: maybe considering passing this parameter by reference, also due to the surface of its usage, maybe defining a type aliasing as `type PayloadList []*ledger.Payload` is more convenient and less error-prone for passing by reference.

File: engine/execution/computation/computer/computer.go
Comment (ramtinms) [reply engine/execution/computation/computer/computer.go:141]: I tried that unfortunately even if we use delta directly, because of `newChild() *View` we have to cast things back again.

File: fvm/state/state_holder.go
Comment (ramtinms) [reply fvm/state/state_holder.go:11]: nice catch, cleaned it.

File: cmd/util/ledger/migrations/accounts.go
Comment (yhassanzadeh13) [inline cmd/util/ledger/migrations/accounts.go:174]:
```suggestion
func (v view) Get(owner, controller, key string) (flow.RegisterValue, error) {
```

File: cmd/util/ledger/migrations/accounts.go
Comment (yhassanzadeh13) [inline cmd/util/ledger/migrations/accounts.go:161]:
```suggestion
		return fmt.Errorf("view type mismatch (given: %T, expected:Delta.View)", o)
```
pretending "cannot merge" to the error is better to be done by the caller.

File: engine/execution/computation/computer/computer.go
Comment (yhassanzadeh13) [inline engine/execution/computation/computer/computer.go:145]:
```suggestion
			return nil, fmt.Errorf("cannot merge view: %w", err)
```

File: engine/execution/computation/computer/computer.go
Comment (yhassanzadeh13) [inline engine/execution/computation/computer/computer.go:178]:
```suggestion
		return nil, fmt.Errorf("cannot merge view: %w", err)
```

File: engine/execution/state/delta/view.go
Comment (yhassanzadeh13) [inline engine/execution/state/delta/view.go:195]:
```suggestion
		return fmt.Errorf("view type mismatch (given: %T, expected:delta.View)", ch)
```

File: engine/execution/state/delta/view.go
Comment (yhassanzadeh13) [inline engine/execution/state/delta/view.go:205]:
```suggestion
		return fmt.Errorf("can not update spock: %w", err)
```

File: engine/execution/state/delta/view_test.go
Comment (yhassanzadeh13) [inline engine/execution/state/delta/view_test.go:229]: shouldn't all these asserts be `require`?

File: fvm/state/address_generator.go
Comment (yhassanzadeh13) [inline fvm/state/address_generator.go:23]: We are always returning a `nil` for this factory. Removing it saves some error checking efforts at the caller side.

File: fvm/state/address_generator.go
Comment (yhassanzadeh13) [inline fvm/state/address_generator.go:28]: I think this method and subsequent ones are not expected to change the underlying `g` receiver. Hence, they may be better to be value receivers instead of pointer receivers.

File: fvm/transactionStorageLimiter_test.go
Comment (yhassanzadeh13) [inline fvm/transactionStorageLimiter_test.go:154]: we have a duplicate implementation of this same function. Maybe it is better to encapsulate it exportable at as common tool.

File: fvm/state/state_test.go
Comment (yhassanzadeh13) [inline fvm/state/state_test.go:38]: shouldn't we also check reading the same key on the parent?

File: fvm/state/state_test.go
Comment (yhassanzadeh13) [inline fvm/state/state_test.go:12]: these tests would be great if supported by documentation. In general, describing what each tries to evaluate.

File: fvm/state/state_test.go
Comment (yhassanzadeh13) [inline fvm/state/state_test.go:59]: I would add reading the value of parent before merge as well.

**Review by yhassanzadeh13 — APPROVED** [review]

File: fvm/state/accounts.go
Comment (ramtinms) [reply fvm/state/accounts.go]: updated with comments

File: fvm/utils/test.go
Comment (ramtinms) [reply fvm/utils/test.go:43]: thats okey, merge view failures cause panics and terminate the nodes.

File: fvm/utils/test.go
Comment (ramtinms) [reply fvm/utils/test.go:70]: actually parent can be nil (very first state and the value might not exist so we return nil, nil.

File: fvm/utils/test.go
Comment (ramtinms) [reply fvm/utils/test.go]: I'll do this in a follow up pr, would be hard to update other PRs depending on this PR right now.

File: fvm/utils/test.go
Comment (ramtinms) [reply fvm/utils/test.go:81]: I'll do this in a follow up pr, would be hard to update other PRs depending on this PR right now.

File: fvm/utils/test.go
Comment (ramtinms) [reply fvm/utils/test.go]: fixed

File: fvm/utils/test.go
Comment (ramtinms) [reply fvm/utils/test.go]: fixed

File: cmd/util/ledger/migrations/accounts.go
Comment (ramtinms) [reply cmd/util/ledger/migrations/accounts.go:140]: I honestly don't think we should spend that much time here, this was a code used few months a code and only kept as a log of operation, we might change the actual logic that happened accidentally.

File: fvm/transactionStorageLimiter_test.go
Comment (ramtinms) [reply fvm/transactionStorageLimiter_test.go:154]: Yeah I intentionally did this to remove the dependency to ledger package, I'll move them into utils in the next PR

File: engine/execution/state/delta/view_test.go
Comment (ramtinms) [reply engine/execution/state/delta/view_test.go:229]: they are both the same, isn't it ?

File: fvm/state/state_test.go
Comment (ramtinms) [reply fvm/state/state_test.go:12]: fixed.

File: fvm/state/state_test.go
Comment (ramtinms) [reply fvm/state/state_test.go:59]: added.

File: fvm/state/state_test.go
Comment (ramtinms) [reply fvm/state/state_test.go:38]: added

File: fvm/state/address_generator.go
Comment (ramtinms) [reply fvm/state/address_generator.go:28]: I honestly didn't touch this code that much, I think we need to change this heavily in an upcoming PR and we might use the error channel.

Comment (ramtinms) [reply fvm/state/address_generator.go:28]: we keep all the interfaces to be pointer in case they need to capture metrics and rate limit number of calls.

**Pull request merged**
